[2단계 - 점심 뭐 먹지] - 상추(이상희) 미션 제출합니다.

sanghee01 opened the pull request:
안녕하세요 도밥!! 이번 리뷰도 잘 부탁드립니다!

리뷰에 앞서 말씀드릴 점이 있는데요! 
사실 step2를 진행하면서 Component Class를 사용하는 부분에서 많은 어려움를 겪었어서, 기능 구현을 다 하지 못했습니다. 

Component에 있는 생명주기 메서드(onRender, template, setState 등)들의 역할이나 호출되는 순서를 제대로 인지하지 못해서 기능 자체는 제대로 구현이 됐지만, 원하는대로 동작하지 않는 경우들이 많았습니다.. 

해당 부분이 해결돼야 기능이 동작할 수 있어서 step2 수행 내내 이 이슈들 해결하는데 시간을 다 쓴 것 같습니다.. 그래서 처음 설정했던 목표와 다른 학습 목표들은 대부분 못 이룬 것 같아 아쉬운 마음이 크네요.. 🥲 
아무튼 이러한 점 때문에, PR 본문에는 거의 해당 내용으로만 채워져 있어, 미리 말씀드리면 좋을 것 같아 말씀드립니다..! 


## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://sanghee01.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 작업 순서

| 작업 내용 | 커밋 |
|---------|------|
| **쓸데없는 class super() 제거** | [🔗 refactor: Class에서 불필요한 생성자 코드 제거 & 불필요한 폴더, 파일 제거](https://github.com/sanghee01/javascript-lunch/commit/d63e5fa4cef1edb42379d800ccc7fe60570c26b9) |
| **모달 select (카테고리, 거리) 컴포넌트화** | [🔗 refactor: modal input select 컴포넌트화](https://github.com/sanghee01/javascript-lunch/commit/ebfc64a725078ec393b84ac841e0cb45a0c3fd00) <br> [🔗 feat: Filter 컴포넌트 추가](https://github.com/sanghee01/javascript-lunch/commit/fd92c9891f21d48fd24e951be369a5c50580f49e) |
| **Prettier 적용** | [🔗 style: prettier 적용](https://github.com/sanghee01/javascript-lunch/commit/ebe24d4a33a3a6c19182a66ff04d8d9295b9c2b9) |
| **Filter 기능 추가** | [🔗 feat: Filter 컴포넌트 추가](https://github.com/sanghee01/javascript-lunch/commit/fd92c9891f21d48fd24e951be369a5c50580f49e) <br> [🔗 feat: 카테고리 filter 기능 추가, sort는 UI만 추가](https://github.com/sanghee01/javascript-lunch/commit/30d024e32afecc2179e0506dfcfd7aba69b4da54) <br> [🔗 fix: 필터 선택 시 지정된 select명으로 안 변했던 문제 수정](https://github.com/sanghee01/javascript-lunch/commit/c9d7ff26b7f6e1f8f13a4954144940f86d46f987) |
| **음식점 이름순, 거리순 정렬 기능 추가** | [🔗 feat: 음식점 목록 이름순, 거리순 정렬 기능](https://github.com/sanghee01/javascript-lunch/commit/99447320d5aef8d2f92dc5331afc00b0d953b7eb) |
| **음식점 디테일 모달창 추가** | [🔗 refactor: 모달 컴포넌트 분리](https://github.com/sanghee01/javascript-lunch/commit/f3f95c2d14901e26bc1f01c44bd578ee4cc1b586) <br> [🔗 feat: 음식점 디테일 모달창 구현 (ESC 키로만 닫힘)](https://github.com/sanghee01/javascript-lunch/commit/82c43d3d0e4f1302bb4b1e8b16168b5290aa343a) <br> [🔗 feat: 음식점 정보에 링크 추가](https://github.com/sanghee01/javascript-lunch/commit/d45cacaf1d5b3f0eda701b743d1c58f54a638792) |
| **로컬스토리지 연동 (음식점 목록 등록)** | [🔗 feat: localStorage 음식점 목록 등록](https://github.com/sanghee01/javascript-lunch/commit/b38e8c0cc8878105a35d5cb502fb3c7b40668a5c) |
| **디테일 모달창 - 삭제 및 닫기 버튼 추가** | [🔗 feat: 모달 디테일 버튼 UI 추가](https://github.com/sanghee01/javascript-lunch/commit/57f90347508fcca581486c05bc73586aa1417647) |
| **음식점 디테일 모달창 닫히는 문제 해결** | [🔗 fix: 디테일 모달 안 닫히는 문제 해결 - 이벤트 위임 사용](https://github.com/sanghee01/javascript-lunch/commit/9f90412cd525c86b9c1c3f97aad912961a7222aa) |
| **TypeScript 간단히 적용** | [🔗 feat: RestaurantType 생성, defaultRestaurantList.ts에 적용](https://github.com/sanghee01/javascript-lunch/commit/f32f14cf8e285da4e3d4cc91093620e6ebc2d753) |

### 구현하지 못한 부분, 개선하고 싶은 부분
- 음식점 삭제 기능(localStorage 포함)
- 즐겨찾기 기능
- TypeScript 가 전반적으로 적용되지 않았음
- E2E Test를 개선하지 못했음
- 재사용 할 수 있을 것 같은 부분 중복 개선하고 싶음(Filter와  Sorter)
- 현재 Restaurant 각각이 모달창을 갖고있는데, 한 곳에서 관리하고 싶음(RestaurantList에서 관리하는거 생각하고 있음)
- 그 외 step1 때의 코멘트

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

앞서 말했듯이, Component Class 자체를 제대로 활용하지 못해 이슈가 많이 생겼어서 대부분 이 부분에 대해 고민했습니다.

- setState 사용시 무한 재귀가 되는 현상이 발생됐습니다. 이는 Filter 컴포넌트에서 발생했던 문제였는데, onRender에서 filterCategory()를 즉시 호출했고, 이 함수가 다시 setState()를 실행하면서 무한 루프가 발생했었습니다.
    - `onRender()`에서 **즉시 `filterCategory()`를 실행** → `setState()` 호출
    - `setState()`가 실행되면 **자동으로 `render()` 실행**
    - `render()`에서는 **다시 `onRender()` 호출**
    => 무한 루프 발생
        
        그래서 `setState()`를 호출하는 대신,  `change` 이벤트 핸들러를 추가하여 이벤트 발생 시 `filterCategory()`를 실행하도록 수정해서 해결했습니다.
        
- 각 Restaurant을 누르면 해당 Restaurant 디테일 모달창이 보이도록 하는 과정에서, 해당 Restaurant을 누르면 아무런 동작이 일어나지 않는 문제를 겪었습니다. 해당 코드는 onRender에 있었는데, onRender 메서드 내부에서 콘솔을 찍어보면 잘 실행이 되는데, 이벤트리스너 부분은 동작하지 않는 문제였습니다.
이는 onRender() 자체는 실행되었지만, 실제로 이벤트가 등록된 요소는 사용되지 않아서 생긴거였습니다. RestaurantList.js에서는 template()만 사용해서 문자열로 변환하고, 그 문자열을 innerHTML로 넣었기 때문에 이벤트가 등록된 기존 요소가 사라지고 새로운 요소가 생깁니다.
그래서 console.log()는 맨처음에 실행되었지만, click 이었던 이벤트 리스너는 동작하지 않았던 것이었습니다
    - 근데 onRender에서 document.addEventListener 로 등록한 이벤트리스너 함수는 잘 동작이 됐습니다. 이유를 탐색한 결과 얘는 전역으로 유지되는거라 어디 위치하든간에 한번 등록되면 쭉 유지되므로 가능하다는 것을 알게됐습니다.
    - 그리고 이 함수가 특정 요소를 리스너하는게 아니라 document를 리스너 하는건데, 원하는대로 동작했던 이유가 궁금해 찾아봤습니다. 이는 이벤트 버블링으로 부모 요소를 가져올 수 있어서 잘 동작됐다는것도 알게됐습니다
- 음식점 디테일 모달 백그라운드나 닫기버튼을 누를때, 콘솔로 찍으면 $modal을 잘 가리키는데 $modal.classList.add('hidden')해도  $modal에 hidden class가 붙지 않는 이슈가 발생했습니다. 그래서 닫혀지지 않았습니다.
    - 이는 이벤트 위임으로 해결 했습니다. 이벤트 위임이 부모에 이벤트 달고 closest로 타고타고 내려가 지정한 것 중 가장 가까운 요소를 잡는다는 것을 알게됐습니다.

이러한 문제를 겪었는데, 크게는 현재 저와 페어가 만든 Component Class를 올바르게 사용하는 방법..? 생명주기..?에 대해 조금 더 알게 되었고, 이벤트 위임에 대해 알게 됐던게 큰 배운점인 것 같습니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

아직 기능 구현도 다 안됐고, 고민하고 개선하고 싶은 포인트들이 많은데 우선순위를 어떻게 둬야할까요..?

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

sanghee01 committed:
feat: 템플릿 복사

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
docs: README 기능 목록 작성

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: 모달 열고닫기

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: 헤더 동적으로 렌더링

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: Component 틀 구현

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: Component 방식 변경 및 메뉴 리스트 렌더링

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: 모달 Component 생성

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: 모달 onRender에서 이벤트리스너 호출

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: 음식점 추가

Co-authored-by: Kuesung Park <gueit214@naver.com>

sanghee01 committed:
feat: input 유효성 검증 추가

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
feat: e2e 테스트 작성

Co-authored-by: Kuesung Park <gueit214@naver.com>

sanghee01 committed:
chore: 안쓰는 파일, 코드 제거

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
refactor: 컴포넌트, css  폴더 분리

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
refactor: 불필요한 주석 제거

Co-authored-by: guesung <gueit214@gachon.ac.kr>

sanghee01 committed:
fix: import 문에서 확장자 붙임. 불필요한 파일 제거

sanghee01 committed:
style: style.css에서 불필요한 코드 제거

sanghee01 committed:
feat: 음식점 카테고리마다 이미지 다르도록 설정

sanghee01 committed:
fix: onRender가 render후에 발생하면 실행되도록 수정

기존에는 상태변경할때 render만 호출돼서 onRender가 실행 안됐음. 그 문제 해결하기 위함

sanghee01 committed:
feat: 음식점 추가시 이미지도 올바르게 변경되도록

sanghee01 committed:
chore: 거리 설명 수정

sanghee01 committed:
feat: 모달 바깥쪽 클릭하거나 ESC 누르면 꺼지도록

sanghee01 committed:
fix: esc 누를때만 모달창 꺼지도록 수정

sanghee01 committed:
refactor: state 변수 private로 설정

sanghee01 committed:
refactor: props private로 변경

sanghee01 committed:
chore: homepage 수정

sanghee01 committed:
refactor: 기본 음식점 데이터들 별도 파일로 분리

sanghee01 committed:
fix: 필수 입력 안할 시 form 추가 안되도록

sanghee01 committed:
fix: 매개변수에 event 객체 넣음

sanghee01 committed:
fix: input label이 정상적으로 동작되도록 수정

sanghee01 committed:
refactor: modal form 태크에 class 추가

sanghee01 committed:
refactor: 불필요한 html 태그 제거

sanghee01 committed:
feat: CategorySelect 컴포넌트 만듦

sanghee01 committed:
refactor: Class에서 불필요한 생성자 코드 제거 & 불필요한 폴더, 파일 제거

헷갈려서 제거

sanghee01 committed:
refactor: modal input select 컴포넌트화

sanghee01 committed:
feat: Filter 컴포넌트 추가

sanghee01 committed:
style: prettier 적용

sanghee01 committed:
feat: 카테고리 filter 기능 추가, sort는 UI만 추가

sanghee01 committed:
fix: 필터 선택시 지정된 select명으로 안변했던거 수정

sanghee01 committed:
feat: 음식점 목록 이름순, 거리순 정렬 기능

sanghee01 committed:
refactor: 모달 컴포넌트 분리

sanghee01 committed:
feat: 음식점 디테일 모달창 구현.. But 백그라운드 클릭시 꺼지진 않음. ESC로 꺼야함

sanghee01 committed:
feat: 음식점 정보에 링크 추가

sanghee01 committed:
feat: localStorage 음식점목록 등록

sanghee01 committed:
feat: 모달 디테일 버튼 UI 추가

sanghee01 committed:
fix: 디테일 모달 안닫히는거 해결 - 이벤트 위임 사용!!!

sanghee01 committed:
feat: RestaurantType 생성, defaultRestaurantList.ts에 적용

sanghee01 committed:
Merge branch 'sanghee01' of https://github.com/woowacourse/javascript-lunch into step2

sanghee01 committed:
fix: 배포본 이미지 경로

sanghee01 committed:
chore: 이미지 경로 수정

sanghee01 committed:
chore: 이미지 경로 수정

Creative-Lee commented on types/types.ts:
```diff
@@ -0,0 +1,8 @@
+  name: string;
+  distance: number;
+  description: string;
+  link: string;
```
? 키워드로 해당 필드가 옵셔널(필수가 아닌) 값이라는 것을 명시할 수 있어요.
위 2개 필드는 필수가 아니니 타입도 맞춰봅시다

Creative-Lee commented on src/components/DetailModal.js:
```diff
@@ -0,0 +1,27 @@
+
+    document.addEventListener('click', function (event) {
+      if (event.target.closest('#modal-cancel')) $modal.classList.add('hidden');
+    });
```
나중에 시간이 난다면 왜 기존 코드에서는 동작하지 않았는지 디버깅 해보는것도 좋은 학습이 될것 같아요 😄 
(해결 과정에서 배운것이 많아서 좋네요 ㅎㅎ)

Creative-Lee commented on src/components/Filter.js:
```diff
@@ -0,0 +1,23 @@
+import Component from '../core/Component';
+
+class Filter extends Component {
```
PR에 적어주신 것 처럼 콜백 함수 prop명을 변경하고 클래스 명을 변경하는등의 몇몇 처리만 한다면, 
재사용 가능한 컴포넌트를 만들 수 있을 것 같네요ㅎㅎ

Creative-Lee commented on src/components/RestaurantList.js:
```diff
@@ -1,22 +1,29 @@
+        if (this.props.sort === '이름순') return a.props.name.localeCompare(b.props.name);
+        if (this.props.sort === '거리순') return a.props.distance - b.props.distance;
+        return 0;
+      });
```
배열 메서드 활용이 좋네요~!
함수 분리까지 한다면 더 좋을것 같습니다 👍

Creative-Lee commented on src/components/InputModal.js:
```diff
@@ -0,0 +1,124 @@
+      new InputBox({
+        input: new Select({
+          name: 'distance',
+          optionList: [5, 10, 15, 20, 30],
```
상수로 관리해볼까요~?

Creative-Lee commented on src/components/Select.js:
```diff
@@ -0,0 +1,16 @@
+import Component from '../core/Component';
+
+class Select extends Component {
```
이 컴포넌트에서도 category와 관련된 내용만 제거한다면 공통컴포넌트로 활용할 수 있겠네요~! 
외부에서 `{ value: xxx, label: xxx }` 형식의 객체 배열(optionList)를 받는다고 가정해서 구현해봅시다. 
(지금의 구현과 크게 달라질 것은 없을거에요 ㅎㅎ)
 
이후에는 `공통 컴포넌트 - Select` 를 활용해서 `도메인 종속적인 컴포넌트 - categorySelect`를 만들 수 있을것 같아요

Creative-Lee reviewed (changes requested):
안녕하세요 상추 👋 

테코톡에 이어 구조적인 문제까지 겪으면서 많이 고생하신것 같네요..😢 
정말 고생많으셨어요~!

작업의 흐름을 적어주셔서 어떤 생각을 가지면서 미션을 진행했는지 파악하기 좋았습니다 😄 
(어떤 역경이 있었는지 간접적으로 느낄 수 있었습니다... 😅)

---

>앞서 말했듯이, Component Class 자체를 제대로 활용하지 못해 이슈가 많이 생겼어서 대부분 이 부분에 대해 고민했습니다.

현재 모든 컴포넌트가 Component class를 기반으로 하기에 거의 앱 전반이 해당 class의 추상화에 의존해 돌아가고 있다고 해도 과언이 아닌데요~! 
겪으신 버그 상황들도 결국 이런 추상화된 로직에 대한 이해가 쉽지 않았기 때문이겠죠.

>이러한 문제를 겪었는데, 크게는 현재 저와 페어가 만든 Component Class를 올바르게 사용하는 방법..? 생명주기..?에 대해 조금 더 알게 되었고, 이벤트 위임에 대해 알게 됐던게 큰 배운점인 것 같습니다.

말씀해주신 것 처럼 오히려 이번에 정말 큰 경험을 하셨다고 생각합니다.
내가 따른 구조에 의해 오히려 발목이 잡히는 경험은 씁쓸하지만 필요한 경험이라는 생각도 듭니다. 
성급하게 로직이나 구조의 추상화를 시도했다가 여러번 데여본 사람이라서 더 공감이 가네요(여전히 그렇기도 하고요 😓)

---
 
마지막 질문에 대해서는 무조건 전자가 우선이라고 생각합니다ㅎㅎ (돌아가는 쓰레기 기억하시죠?)
현업에서도 개선포인트가 잔뜩임에도 눈물을 머금고 기능구현을 하게되는 상황이 많습니다.
그래서 중간중간 습관적으로 리팩터링 해주는게 중요하다고 생각하기도 하고요. 

필요한 부분 채워서 요청주시면 되겠습니다~!
정말 고생많으셨어요 상추💪

sanghee01 committed:
refactor: 필수 데이터가 아닌 값 ts 옵셔널 값 지정

sanghee01 committed:
feat: localStorage 삭제 기능 추가, 음식점 디테일모달 리팩토링

리팩토링 내용: 각 Restaurant에서 각 모달을 띄우도록 했는데, RestaurantList 한곳에서만 DetailModal을 생성하고 띄우도록 함.

localStorage 삭제 기능: 각 음식점을 인식하기 위해 data 값 추가. Application에서 삭제 메서드 만듦

sanghee01 committed:
refactor: Component 생성시 template로 요소 생성

div가 겹겹이 있는 걸 없애기 위함

sanghee01 committed:
refactor: gnbButton 위치 Header로 이동, Modal 중복된 코드 리팩토링

sanghee01 committed:
refactor: InputBoxList 파일 분리

sanghee01 committed:
refactor: Filter, Sorter 하나의 컴포넌트로

sanghee01 committed:
feat: Restaurant 즐겨찾기 UI

sanghee01 committed:
feat: 음식점 즐겨찾기 기능, localStorage 연동

sanghee01 committed:
fix: DetailModal에서 즐겨찾기 기능 수행시, 부모의 즐찾 기능 수행 하지 못하도록 설정 (이벤트 버블링 방지)

sanghee01 committed:
feat: 즐겨찾기 탭

sanghee01 committed:
refactor: 공통 컴포넌트 폴더 분리

sanghee01 committed:
refactor: 컴포넌트  export index.js 에서 관리

sanghee01 committed:
refactor: RestaurantList 로직 함수 분리

sanghee01 committed:
refactor: DetailModal 불필요한 코드 제거

sanghee01 committed:
refactor: input 데이터 FormData API로 불러오도록 / 함수분리

sanghee01 committed:
refactor: 음식점 입력 부분 상수화

sanghee01 committed:
chore: 이미지 경로 수정

sanghee01 committed:
chore: 이미지 경로 수정

sanghee01 committed:
chore: 이미지 경로 수정

sanghee01 committed:
chore: 이미지 경로 수정

sanghee01 committed:
docs: Update REAMDE

sanghee01 commented:
안녕하세요 도밥! 기능 요구사항에 맞게 구현 다 완료했습니다! 코멘트에 남긴 부분도 어느정도 반영했습니다! 

### 가장 고민했던 부분
이번에 미션을 구현, 리뷰 반영하면서 각 컴포넌트의 역할, 책임에 대해 고민할 수 있었습니다. 예를들어, input Modal에 gnb header 관련 코드가 있는게 이상하다는걸 느꼈습니다. 해당 코드가 자기의 컴포넌트의 책임에 맞는 코드인가?를 많이 고민했던 것 같습니다.

### 가장 큰 고난을 겪은 부분
이벤트 위임을 사용함으로써 이벤트 버블링, 이벤트 전파 관련 문제도 겪은 것입니다.. 이벤트 위임으로 document에서 event를 Listen하니까, 동일한 이벤트함수를 쓰는 곳에서 이상하게 작동하더라고요. 
예를들어, 현재 즐겨찾기 기능이 메인 페이지, 디테일모달 두군데에서 수행할 수 있어서 발생한 문제가 있었습니다. DetailModal이 RestaurantList의 DOM 계층 내에 삽입되어 이벤트 버블링이 발생해, 즐겨찾기 클릭 이벤트가 두 번 처리되면서, DetailModal 내부의 즐겨찾기 이미지 클릭 시에도 RestaurantList의 이벤트 리스너가 실행되는 문제 발생했습니다.

이벤트 버블링 관련된 문제인 것 같은데, 일단 event.stopPropagation()하고, 상위 이벤트 핸들러에서 조건 검사하는 식으로 해결했습니다. 더 깊게 원인과 해결방안을 찾고 싶었으나 기능구현을 해야하니 더 깊게 찾아보진 못했습니다.. 근데 나중에 또 겪게될 것 같아서 주말에 정리하려고 합니다..! 

### 아쉬운 부분
#### 디테일 모달에서의 즐겨찾기
디테일 모달에서 즐겨찾기를 누르면 모달창이 꺼집니다. 이 이유를 찾아보니, 즐겨찾기 여부를 변경할때 setState를 사용함으로써 리렌더가 되니까 꺼지는 것 같습니다. 근데 setState를 실행하면 render를 실행할 수 밖에 없어서(바로 반영해야하니까) 달리 수정할 방법이 떠오르지 않았습니다. 이걸 위해 부모 클래스인 Component 내부의 setState를 수정하기엔.. 다른 자식 클래스에도 영향이 갈 것 같아서 그러지 않았습니다. 어떻게 해야할지 모르겠고 일단 기능요구사항에는 맞으니 그냥 냅두었습니다 ㅠ.ㅠ

#### 도밥이 주신 고민포인트, 팁
그 외에도 시간상 도밥이 제시해주신 고민포인트들 중 몇가지 적용 못한 부분이 있습니다. 그래도 gpt를 통해 왜 그렇게 수정해야하는지는 알게되었습니다. 지금 상황에서 그것까지 적용하기엔 시간이 부족하고 코드 구조를 좀 바꿔야하는 경우도 있어서 그냥 이런 방법이 있구나!! 정도로만 이해했습니다.
 
1. '#app에 dom이 없다면 어떻게 될까요?' 하는 부분에는 만약 #app 요소가 없다면 querySelector는 null을 반환하고, 그 후에 appendChild를 호출할 때 TypeError가 발생하는 문제가 있음을 알게 됐습니다. 
TS 연관해서는 TS에서는 null 가능성을 명시적으로 처리해야 하는데, non-null assertion 을 사용하면 런타임 에러가 발생할 수 있으므로, DOM 요소가 존재하는지 확인하는 것이 중요하다는 것을 알게됐습니다.

2. DOM 선택 메서드
DOM util ($, $$)사용하면 확실히 깔끔해질 것 같다는 생각이 드네요.  다음에 기회가 된다면 사용하고 싶다는 생각이 들었습니다. 뭔가 다 변환하기엔 우선순위에 미뤄져있어서 적용하진 못했습니다.
이 유틸리티 함수를 통해 앞서 말한 1번 포인트(#app 부분)부분을 확장해서, 아래 코드와같이 각 컴포넌트를 null check 하는 코드도 넣으면 더 유용히 사용할 수 있을 것 같다는 생각이 들었습니다.
```javascript
export const $ = (selector, parent = document) => {
  const element = parent.querySelector(selector);
  if (!element) {
    console.error(`Error: Selector "${selector}" did not match any element.`);
  }
  return element;
};
```

#### ts, e2e
아무래도 기능 구현에 집중하다보니 ts와 e2e는 개선을 못한점이 아쉽네요. 이는 영화리뷰 미션을 수행하면서 더 학습해볼까 합니다! 

### 그 외에 알게된 점
1. index.js에서 export {default as ~ } from ~를 사용한 이유
이렇게 함으로써 여러 컴포넌트를 중앙에서(index.js) 관리하고, 다른 모듈이 보다 쉽게 접근할 수 있다는 장점으로 사용함을 알게되었습니다. import를 간결히 작성할 수 있고, 새로운 컴포넌트를 추가하거나 경로가 변경될 때, index.js만 수정하면 되므로 유지보수가 편해지므로 모듈 관리에도 용이하다는 것을 알게되었습니다. 

2. input 값 쉽게 가져오는 방법: FormData API
HTMLFormElement의 FormData API를 활용하면 한 번에 모든 입력값을 가져올 수 있다는것을 알게되었습니다. 이를 사용하면 form 요소에 있는 모든 입력값을 key-value 쌍으로 쉽게 가져올 수 있고, input, select, textarea의 name속성으로 key를 사용할 수 있음을 알게됐습니다. 또한 form 요소는 반드시 `<form>`태그여야하는것도 알게됐습니다.

이걸 사용하면 개별로 각 input을 찾아서 값을 추출하는 코드가 필요 없다는 것과, 만약 form 내부의 input 요소가 추가되거나 이름이 바뀌더라도, FormData는 자동으로 해당 입력값을 가져오므로 유지보수에 좋다는것을 알게됐습니다.

3. DocumentFragment
중복되는 div를 template로 어느정도 해결했습니다. 이는 fragment역할을 하는데, innerHTML으로 내부에 값을 넣어도 XSS 공격을 받지 않다는 것을 알게됐습니다. 다만 시간상 더 못찾아본게 아쉬워서 이것도 추후 더 찾아보려고 합니다.

### 작업 순서
| 작업 내용 | 커밋 |
| --- | --- |
| TS 필수 데이터 아닌 값 옵셔널 값 지정 | [refactor: 필수 데이터가 아닌 값 ts 옵셔널 값 지정](https://github.com/woowacourse/javascript-lunch/pull/263/commits/a479f6cf20ca596334475c2dd74e272c5c1e9411) |
| 로컬스토리지 삭제 기능 구현, Restaurant 각각이 모달창을 갖던 것을 그냥 RestaurantList에서 하나에서 관리 | [feat: localStorage 삭제 기능 추가, 음식점 디테일모달 리팩토링](https://github.com/woowacourse/javascript-lunch/pull/263/commits/02fb97adc9d50859e4563f2133f60b4bfbe0e847) |
| render시 불필요하게 div 생성됐던거 수정 | [refactor: Component 생성시 template로 요소 생성](https://github.com/woowacourse/javascript-lunch/pull/263/commits/1cd9c10276251f26e5b4e94f89dffda2b7a054d2) |
| inputModal 책임 이상한거 수정 | [gnbButton 위치 Header로 이동, Modal 중복된 코드 리팩토링](https://github.com/woowacourse/javascript-lunch/pull/263/commits/0d7550dad0fb4bcc06ff84bf37463fd1973ce939)<br>[refactor: InputBoxList 파일 분리](https://github.com/woowacourse/javascript-lunch/pull/263/commits/b7dfc4af87290abf95f63a23fe31f7128c91d9f1) |
| Filter, Sorter 재사용 | [refactor: Filter, Sorter 하나의 컴포넌트로](https://github.com/woowacourse/javascript-lunch/pull/263/commits/d305129131117cc8b3005a19584a2634b2fe0abc) |
| 즐겨찾기 기능 구현 | [feat: Restaurant 즐겨찾기 UI](https://github.com/woowacourse/javascript-lunch/pull/263/commits/a6b8d097ffb4fa0f01616c27b10bdb0fca8f7719)<br>[feat: 음식점 즐겨찾기 기능, localStorage 연동](https://github.com/woowacourse/javascript-lunch/pull/263/commits/7ee9728ebb1d478bdb6f8f4f3aead51ff7091622)<br>[fix: DetailModal에서 즐겨찾기 기능 수행시, 부모의 즐찾 기능 수행 하지 못하도록 설정 (이벤트 버블링 방지)](https://github.com/woowacourse/javascript-lunch/pull/263/commits/fc1f9d528c1d1d51b94444bc08242404254265fb) |
| 즐겨찾기 탭 구현 | [feat: 즐겨찾기 탭](https://github.com/woowacourse/javascript-lunch/pull/263/commits/bdbff0687f64b814fd7f4ed2a6e5c62e7df5d33c) |
| InputModal 리팩토링 - 설정 관련 코드 파일 분리 | [refactor: InputBoxList 파일 분리](https://github.com/woowacourse/javascript-lunch/pull/263/commits/b7dfc4af87290abf95f63a23fe31f7128c91d9f1) |
| 공통 컴포넌트 폴더 분리 | [refactor: 공통 컴포넌트 폴더 분리](https://github.com/woowacourse/javascript-lunch/pull/263/commits/bb7421c6d95d9fd3fb218c59d4f167197e47c016) |
| 컴포넌트 export index.js에서 관리 | [refactor: 컴포넌트 export index.js 에서 관리](https://github.com/woowacourse/javascript-lunch/pull/263/commits/b644791194e3660e08f9b50f4a63bc14ff76947f) |
| RestaurantList 로직 함수 분리 | [refactor: RestaurantList 로직 함수 분리](https://github.com/woowacourse/javascript-lunch/pull/263/commits/eb6219c0ff301d96cf4b0e1e25d705c28cc21f92) |
| DetailModal 불필요한 코드 제거 | [refactor: DetailModal 불필요한 코드 제거](https://github.com/woowacourse/javascript-lunch/pull/263/commits/5cf5f6149354209d07b11f332f6b45f4104ecf79) |
| input 데이터 FormData API로 가져오도록 & 함수 분리 | [refactor: input 데이터 FormData API로 불러오도록 / 함수분리](https://github.com/woowacourse/javascript-lunch/pull/263/commits/e53aa58aebb74dce89e2972b566f828275ea168f) |
| InputModal 상수 관리 | [refactor: 음식점 입력 부분 상수화](https://github.com/woowacourse/javascript-lunch/pull/263/commits/a94135b7ceb8b91dbbbced5a0cf63929f059c402) |
| GitHub Pages 배포 이미지 경로 수정 | [chore: 이미지 경로 수정](https://github.com/woowacourse/javascript-lunch/pull/263/commits/cda2f0cd824d2cee3eb2c72a9d48e0882091dbe8) |

Creative-Lee commented on types/types.ts:
```diff
@@ -3,6 +3,6 @@ export interface RestaurantType {
   distance: number;
-  description: string;
-  link: string;
+  description?: string;
+  link?: string;
```
👍

Creative-Lee commented on src/components/Modal.js:
```diff
@@ -3,13 +3,27 @@ import Component from '../core/Component';
+
+    document.addEventListener('keydown', (event) => {
+      if (event.key === 'Escape') this.element.classList.add('hidden');
+    });
```
모든 이벤트를 document에 부착하는 구현은 직접 겪으신것 처럼 
이벤트 전파와 관련된 문제도 존재하고, 유지보수도 힘들다고 생각해요.
이벤트가 몇개만 추가되어도 관련되어 얽히고 얽힌 문제들이 마구 쏟아질겁니다~!

모달과 관련된 이벤트를 모달컴포넌트와 가깝게 위치시킨것에서 더 나아가, 이벤트도 관련 DOM에 부착하여 관리하면 좋을것 같습니다 ㅎㅎ

Creative-Lee commented on src/components/RestaurantList.js:
```diff
@@ -23,6 +23,35 @@ class RestaurantList extends Component {
     });
+
+    this.element.addEventListener('click', (event) => {
+      const restaurantElement = event.target.closest('.restaurant');
```
`closest`는 DOM구조 변경에 취약하다는 단점도 존재해요. 
DOM 구조는 생각보다 쉽게 변경될 수 있기에 이런 상황도 고려해보면 좋겠습니다 ㅎㅎ

Creative-Lee commented on src/components/InputModal.js:
```diff
@@ -0,0 +1,61 @@
+
+class InputModal extends Component {
+  getModalInput(modalForm) {
+    const baseURL = window.location.origin.includes('github.io') ? '' : '../../public/images/';
```
이미지 경로관련 설정에 대해서는 다른 크루들 코드를 참고해봐도 좋겠어요 ㅎㅎ
(개인의 생각이 반영되는 코드는 아니니까요~!)

Creative-Lee commented on src/components/RestaurantList.js:
```diff
@@ -1,22 +1,93 @@
+      }
+    });
+
+    this.element.addEventListener('click', (event) => {
+      if (event.target.closest('.modal')) return;
+
+      if (event.target.closest('.restaurant__favorite')) {
+        const restaurantElement = event.target.closest('.restaurant');
+        if (event.target.src.includes('filled')) {
+          event.target.src = event.target.src.replace('filled', 'lined');
+          this.props.removeFavorite(restaurantElement.dataset.id);
+        } else {
+          event.target.src = event.target.src.replace('lined', 'filled');
+          this.props.addFavorite(restaurantElement.dataset.id);
+        }
+      }
+      event.stopPropagation();
+    });
```
이 코드라인에서 '레스토랑 리스트 컴포넌트의 책임이 어디까지인가' 를 고민해볼 수 있을것 같습니다~!

Creative-Lee commented on src/components/index.js:
이런 파일 정리 패턴에 장점도 있지만, 단점도 있답니다. 
항상 장단점을 세트로 학습해보시면 좋겠습니다 ㅎㅎ

Creative-Lee reviewed (approved):
안녕하세요 상추😄 
PR에 적어주신 내용만 한트럭이네요~!! ㅎㅎㅎ 
그만큼 삽질한것도, 배운것도 엄청나게 많은 미션이었음을 느낄 수 있었습니다.


>가장 고민했던 부분
이번에 미션을 구현, 리뷰 반영하면서 각 컴포넌트의 역할, 책임에 대해 고민할 수 있었습니다. 예를들어, input Modal에 gnb header 관련 코드가 있는게 이상하다는걸 느꼈습니다. 해당 코드가 자기의 컴포넌트의 책임에 맞는 코드인가?를 많이 고민했던 것 같습니다.

너무 훌륭합니다👍  이번 미션의 전부라고 해도 과언이 아닐거에요.
컴포넌트의 책임, 컴포넌트의 역할, 컴포넌트가 받아야할 인자.... 등 컴포넌트 자체에 대한 중요한 생각들을 잔뜩 하셨던것 같습니다.
지금 진행하고 계신 영화미션에서, 나아가 레벨2에서도 동일합니다. 
하루종일 고민하고 생각할만한 가치가 있는 고민들이에요~! 고생하셨습니다!

>가장 큰 고난을 겪은 부분
이벤트 위임을 사용함으로써 이벤트 버블링, 이벤트 전파 관련 문제도 겪은 것입니다.. 이벤트 위임으로 document에서...

좋은 경험이라고 생각합니다. 
상위에 모든 이벤트를 몰아서 부착했을 때는 어떤 장단이 있는지 확실하게 느끼지 않았을까 싶어요.
말씀해주신 것 처럼 이벤트 전파와 관련되어 겪었던 문제에 대해 반드시 복기해보시길 바라요.
싹 다 상추것으로 만들어보세요💪 
(요구 사항 구현을 우선시 해주신것은 특히 칭찬드리고 싶습니다😄 - 앱은 돌아가는게 중요합니다. 항상 잊지 마셔요!)

>아쉬운 부분
생략...
이걸 위해 부모 클래스인 Component 내부의 setState를 수정하기엔.. 다른 자식 클래스에도 영향이 갈 것 같아서 그러지 않았습니다.

이미 앱 전반에 자리잡은 구조로 인해 특정 부분을 수정하기 어려운 상황을 겪은것은 상추에게 큰 경험이 될거에요. 
그만큼 초기 설계 단계에서는 구조 추상화 같은 부분에서 조심스럽게 접근해야한다는 것을 배웠을 것 같습니다.
내가 작성한 코드, 선택한 구조가 현재 상황에서 정말 필요한게 맞는지 항상 다시 체크해보는 습관을 길러봅시다~! (스스로 납득되며 충분히 이해해야하고, 상대방을 설득할 근거가 있어야합니다.)

> 도밥이 주신 고민포인트, 팁 및 그 외에 알게된 점

바쁜 와중에 상당히 많이 학습하셨네요😮  대단합니다👍 
이번 미션에서 배운것들만 확실하게 복습해도 정말 큰 도움이 될거라고 장담합니다 💯 

---

요구사항 모두 만족해서 구현된것 확인했습니다. (앱 잘 돌아가네요👍)
테코톡에 미션2개까지 겹쳐서 지금까지 진행한다고 힘드셨을텐데 대단합니다.
주말에 복기해보며 '와 이렇게 까지 많이 배웠다고?' 생각이 절로 들 것 같네요💪 
남은 영화미션에서도 지금처럼만 학습 잘 해주시면 좋겠습니다~! 
화이팅하시고 레벨1 마무리 잘하셔요 😄 정말 고생많으셨습니다 상추💯
